plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/RunTimeDependencyValidator.xtend

```diff
@@ -0,0 +1,247 @@
+import org.eclipse.n4js.validation.IssueCodes
+import org.eclipse.n4js.validation.JavaScriptVariantHelper
+import org.eclipse.xtext.EcoreUtil2
+import org.eclipse.xtext.validation.Check
+import org.eclipse.xtext.validation.EValidatorRegistrar
+
+import static org.eclipse.n4js.utils.N4JSLanguageUtils.*
+
+/**
+ * Validations related to run-time dependencies (in particular, illegal load-time dependency cycles).
+ * This class relies on the TModule-information computed in {@link RunTimeDependencyProcessor}.
+ */
+class RunTimeDependencyValidator extends AbstractN4JSDeclarativeValidator {
+
+	private static final String INDENT = "    ";
+
+	@Inject
+	private JavaScriptVariantHelper javaScriptVariantHelper;
+
+	/**
+	 * NEEDED
+	 *
+	 * when removed check methods will be called twice once by N4JSValidator, and once by
+	 * AbstractDeclarativeN4JSValidator
+	 */
+	override register(EValidatorRegistrar registrar) {
+		// nop
+	}
+
+	/**
+	 * A non-inheritance load-time reference R from a module m to an identifiable element in module m’ is an error iff
+	 * <ul>
+	 * <li>case m != m’: m, m’ are run-time cyclic to each other,
+	 * <li>case m = m’: m is run-time cyclic to some other module m’’.
+	 * </ul>
+	 */
+	@Check
+	def void checkIllegalLoadTimeReference(IdentifierRef idRef) {
+		if (!N4JSASTUtils.isTopLevelCode(idRef)) {
+			return; // only interested in top-level == load-time references here!
+		}
+		val targetModule = getTargetModule(idRef);
+		if (targetModule === null) {
+			return;
+		}
+		val containingModule = (idRef.eResource as N4JSResource).module;
+		val hasCycle = (targetModule === containingModule && !containingModule.cyclicModulesRunTime.empty)
+			|| (targetModule !== containingModule && containingModule.cyclicModulesRunTime.contains(targetModule)); // FIXME linear search
+		if (hasCycle) {
+			val message = IssueCodes.getMessageForLTD_ILLEGAL_LOAD_TIME_REFERENCE()
+				+ '\n' + dependencyCycleToString(containingModule, false, INDENT);
+			addIssue(message, idRef, IssueCodes.LTD_ILLEGAL_LOAD_TIME_REFERENCE);
+		}
+	}
+
+	@Check
+	def void checkIllegalImportOfLTSlave(Script script) {
+		val importDecls = script.scriptElements.filter(ImportDeclaration).toList;
+
+		val modulesInHealedCycles = new HashSet<TModule>();
+		for (importDecl : importDecls) {
+			val containingModule = (importDecl.eResource as N4JSResource).module;
+
+			if (holdsNotAnIllegalImportWithinLoadTimeCycle(containingModule, importDecl)
+				&& holdsNotAnIllegalImportOfLTSlave(containingModule, importDecl, modulesInHealedCycles)) {
+
+				// we have a valid import, it may contribute to healing later imports:
+				val targetModule = importDecl.module;
+				if (!targetModule.cyclicModulesRunTime.empty) {
+					modulesInHealedCycles += targetModule;
+					modulesInHealedCycles += targetModule.cyclicModulesRunTime;
+				}
+			}
+		}
+	}
+
+	def private boolean holdsNotAnIllegalImportWithinLoadTimeCycle(TModule containingModule, ImportDeclaration importDecl) {
```

Comment:
Maybe you could add a comment here like at the other methods... just to be complete.